src/hotspot/cpu/x86/macroAssembler_x86.cpp

```diff
@@ -5591,7 +5592,8 @@ void  MacroAssembler::decode_klass_not_null(Register r, Register tmp) {
     shlq(r, LogKlassAlignmentInBytes);
   }
   if (CompressedKlassPointers::base() != nullptr) {
-    mov64(tmp, (int64_t)CompressedKlassPointers::base());
+    // Uses 32-bit mov if base is small enough
+    movptr(tmp, (intptr_t)CompressedKlassPointers::base());
```

Comment:
If we can reserve the base in the low 2G then this could be optimized further to `addq(r, CompressedKlassPointers::base())` and if `LogKlassAlignmentInBytes` is 8 (as asserted in `decode_and_move_klass_not_null` below), we can shorten the whole sequence into `leaq(r, Address(noreg, r, Address::times_8, CompressedKlassPointers::base())`

Reply:
Nice. But could we do this in a follow-up RFE please? I don't have any more time to spend on this atm, and I would like to get this PR at least integrated. It has been sitting in review limbo for 7 weeks now.

I opened a follow-up RFE: https://bugs.openjdk.org/browse/JDK-8326383

(about the leaq, we may move to a different encoding scheme with a different shift value with Lilliput, so I am hesitant to put in too much work that relies on a 8-byte shift)

Reply:
I agree with that.